rust/exchange_rate/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+                    None => {
+                        // The requested time rate isn't found in map. Send a canister get_rate call to self
+                        tokio::spawn(async move { get_rate(valid_job).await });
+                    }
+                } );
+            }
+            Err(weird_job) => {
+                println!("Invalid job found in the request queue! The job Id should be a Unix timestamp divided by 60, e.g., represents a timestamp rounded to minute. Wrong Job Id: {weird_job}");
+            }
+        }
+    });
+
+    return;
+}
+
+/*
+A function to call IC http_request function with a single minute range.
+This function is to be triggered by timer as jobs move to the tip of the queue.
+ */
+#[allow(dead_code)]
+async fn get_rate(key: Timestamp) {
+    let start_time = key * 1000;
+    let end_time = (key + 1) * 1000 - 1;
+
+    // prepare system http_request call
+    let mut request_headers = vec![];
+    request_headers.insert(
+        0,
+        HttpHeader {
+            name: "Connection".to_string(),
+            value: "keep-alive".to_string(),
+        },
+    );
+
+    let request = CanisterHttpRequestArgs {
+        url: format!("https://api.binance.com/api/v3/klines?symbol=ICPUSDT&interval=1m&startTime={start_time}&endTime={end_time}"),
+        http_method: HttpMethod::GET,
+        body: None,
+        transform_method_name: Some("sanitize_response".to_string()),
+        headers: request_headers,
+    };
+
+    let body = candid::utils::encode_one(&request).unwrap();
+
+    match ic_cdk::api::call::call_raw(
+        Principal::management_canister(),
+        "http_request",
+        &body[..],
+        0,
+    )
+    .await
+    {
+        Ok(result) => {
+            // decode the result
+            let decoded_result = candid::utils::decode_one(&result).unwrap();
+
+            // put the result to hashmap
+            FETCHED.with(|fetched_lock| {
+                let mut stored = fetched_lock.borrow_mut();
+                stored.insert(key, decoded_result);
+            });
+        }
+        Err((r, m)) => {
+            println!("The http_request resulted into error. RejectionCode: {r:?}, Error: {m}");
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[candid_method(query, rename = "sanitize_response")]
+async fn sanitize_response(
+    raw: Result<CanisterHttpResponsePayload, Error>,
+) -> Result<CanisterHttpResponsePayload, Error> {
+    match raw {
+        Ok(mut r) => RESPONSE_HEADERS_SANTIZATION.with(|response_headers_blacklist| {
+            let mut processed_headers = vec![];
+            for header in r.headers.iter() {
+                if !response_headers_blacklist.contains(&header.name.as_str()) {
+                    processed_headers.insert(0, header.clone());
+                }
+            }
+            r.headers = processed_headers;
+            return Ok(r);
+        }),
+        Err(m) => {
+            return Err(m);
+        }
+    }
+}
+
+#[allow(dead_code)]
+fn main() {}
```

Comment:
See https://github.com/Psychedelic/DIP20/blob/9053e310f0e82d0251f786ae3f69a0d8961a2972/rust/token/src/main.rs#L629 for a good trick to see the candid.

Reply:
Adopted suggestion.